File: Go/internal/service/service.go
package service

import (
	"errors"
	"magnetar-go/internal/model"
	"magnetar-go/internal/repository"
)

var (
	ErrUserNotFound = errors.New("user not found")
	ErrTaskNotFound = errors.New("task not found")
)

type Service struct {
	repo repository.Repository
}

func NewService(repo repository.Repository) *Service {
	return &Service{repo: repo}
}

func (s *Service) CreateUser(user *model.User) error {
	return s.repo.CreateUser(user)
}

func (s *Service) GetUser(id uint) (*model.User, error) {
	return s.repo.GetUser(id)
}

func (s *Service) GetAllUsers() ([]model.User, error) {
	return s.repo.GetAllUsers()
}

func (s *Service) UpdateUser(id uint, updates map[string]interface{}) (*model.User, error) {
	user, err := s.repo.GetUser(id)
	if err != nil {
		return nil, err
	}

	if name, ok := updates["name"].(string); ok {
		user.Name = name
	}
	if email, ok := updates["email"].(string); ok {
		user.Email = email
	}

	if err := s.repo.UpdateUser(user); err != nil {
		return nil, err
	}
	return user, nil
}

func (s *Service) DeleteUser(id uint) error {
	_, err := s.repo.GetUser(id)
	if err != nil {
		return err
	}
	return s.repo.DeleteUser(id)
}

func (s *Service) CreateTask(userID uint, task *model.Task) error {
	// Verify user exists
	_, err := s.repo.GetUser(userID)
	if err != nil {
		return ErrUserNotFound
	}
	task.UserID = userID
	return s.repo.CreateTask(task)
}

func (s *Service) GetTask(id uint) (*model.Task, error) {
	return s.repo.GetTask(id)
}

func (s *Service) GetAllTasksByUserID(userID uint) ([]model.Task, error) {
	// Verify user exists
	_, err := s.repo.GetUser(userID)
	if err != nil {
		return nil, ErrUserNotFound
	}
	return s.repo.GetAllTasksByUserID(userID)
}

func (s *Service) UpdateTask(id uint, updates map[string]interface{}) (*model.Task, error) {
	task, err := s.repo.GetTask(id)
	if err != nil {
		return nil, err
	}

	if title, ok := updates["title"].(string); ok {
		task.Title = title
	}
	if description, ok := updates["description"].(string); ok {
		task.Description = description
	}

	if err := s.repo.UpdateTask(task); err != nil {
		return nil, err
	}
	return task, nil
}

func (s *Service) MarkTaskDone(id uint) (*model.Task, error) {
	task, err := s.repo.GetTask(id)
	if err != nil {
		return nil, err
	}
	task.Done = true
	if err := s.repo.UpdateTask(task); err != nil {
		return nil, err
	}
	return task, nil
}

func (s *Service) DeleteTask(id uint) error {
	_, err := s.repo.GetTask(id)
	if err != nil {
		return err
	}
	return s.repo.DeleteTask(id)
}

File: Zig/src/service.zig
const std = @import("std");
const models = @import("models.zig");
const Repository = @import("repository.zig").Repository;

pub const Service = struct {
    repo: *Repository,
    allocator: std.mem.Allocator,

    pub fn init(repo: *Repository, allocator: std.mem.Allocator) Service {
        return Service{
            .repo = repo,
            .allocator = allocator,
        };
    }

    // Users
    pub fn createUser(self: *Service, name: []const u8, email: []const u8) !models.User {
        if (name.len == 0 or email.len == 0) return error.ValidationFailed;

        const user = models.User{
            .name = name,
            .email = email,
        };
        return self.repo.createUser(user);
    }

    pub fn getUser(self: *Service, id: i64) !models.User {
        if (try self.repo.getUser(id, self.allocator)) |user| {
            return user;
        }
        return error.NotFound;
    }

    pub fn getAllUsers(self: *Service) ![]models.User {
        return self.repo.getAllUsers(self.allocator);
    }

    pub fn updateUser(self: *Service, id: i64, name: ?[]const u8, email: ?[]const u8) !void {
        if (name == null and email == null) return; // Nothing to update
        try self.repo.updateUser(id, name, email);
    }

    pub fn deleteUser(self: *Service, id: i64) !void {
        try self.repo.deleteUser(id);
    }

    // Tasks
    pub fn createTask(self: *Service, user_id: i64, title: []const u8, description: []const u8) !models.Task {
        if (title.len == 0) return error.ValidationFailed;

        // Verify user exists
        if (try self.repo.getUser(user_id, self.allocator)) |user| {
             self.allocator.free(user.name);
             self.allocator.free(user.email);
        } else {
            return error.NotFound; // User not found
        }

        const task = models.Task{
            .user_id = user_id,
            .title = title,
            .description = description,
            .done = false,
        };
        return self.repo.createTask(task);
    }

    pub fn getTask(self: *Service, id: i64) !models.Task {
         if (try self.repo.getTask(id, self.allocator)) |task| {
            return task;
        }
        return error.NotFound;
    }

    pub fn getUserTasks(self: *Service, user_id: i64) ![]models.Task {
        // Verify user exists first? Maybe not strictly required by spec but good practice.
        // Spec says GET /users/{id}/tasks, implies 404 if user doesn't exist?
        // Or just empty list. Let's return tasks.
        return self.repo.getUserTasks(user_id, self.allocator);
    }

    pub fn updateTask(self: *Service, id: i64, title: ?[]const u8, description: ?[]const u8, done: ?bool) !void {
        try self.repo.updateTask(id, title, description, done);
    }

    pub fn markTaskDone(self: *Service, id: i64) !void {
        try self.repo.markTaskDone(id);
    }

    pub fn deleteTask(self: *Service, id: i64) !void {
        try self.repo.deleteTask(id);
    }
};
